### ISSUE TITLE
Index explorer views to make them findable by external search

### ISSUE THREAD

**Issue by larsyencken** [owid/owid-grapher#1214]
## Core problem

We are putting more and more charts into explorers, but only graphers are really indexed well. This means if you only have a chart in an explorer, you will never find it with Google search.

## Proposed solution

To avoid having to make a duplicate grapher chart just for SEO, we should index each view of each explorer in our sitemap.

## Challenges

* We have both very curated explorers, and massive "reference" explorers with thousands of views. They don't quite seem to be equally relevant. **Do we need a way to select relevant views?**
* Some of our explorers have different views that have the same title; e.g. the demography explorer doesn't have a different title for the different projections. **Will this cause problems or confusion?**
* **How do we measure success?**
* **How, and when, will Google surface these as results?**
* Our pages are JS-only, and without running a lot of JS Google cannot infer the chart title, subtitle, or most other relevant metadata. **Will this cause problems for us?**

**Comment by danielgavrilov** [comment]
This one might need some research, some back and forth experiments to figure out how we can "teach" Google to index client-side views separately.

**Comment by larsyencken** [comment]
Scheduling this one for @marcelgerber to investigate this cycle.

**Comment by marcelgerber** [comment]
Progress update:

### What I've learned already

To my surprise, Google really does render out the full explorer every time it's crawling our site.
We can see this in Google Search Console by inspecting a URL - e.g. https://ourworldindata.org/explorers/population-and-demography - then clicking on "View crawled page" and observing that the crawled page includes the rendered-out chart, chart title etc., which can only be filled in once the data file has been fetched.

### What I've changed

At this point, I've already [fixed the canonicalUrl for explorer pages](https://github.com/owid/owid-grapher/commit/e85c1016e25746f4c2e73a635d61ed3c9957e9f8) (which was wrong before) and [included explorers in our sitemap](https://github.com/owid/owid-grapher/pull/1641).
Both of these should make Google a little bit happier with indexing our explorers.

### What's still puzzling

So, currently we have a `canonicalUrl` set on every explorer page that points to the default view (without any params). This seems sensible, but _might_ prevent Google from indexing any other view at all.
In [their help page on canonical URLs](https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls), Google says:
> The canonical page will be crawled most regularly; duplicates are crawled less frequently in order to reduce the crawling load on your site.

They don't explicitly say that any other page will never be indexed, but that seems to be the behaviour.
For example, searching for https://ourworldindata.org/explorers/population-and-demography?facet=none&hideControls=true&Metric=Fertility+rate&Sex=Both+sexes&Age+group=Total&Projection+Scenario=Medium&country=High-income+countries~OWID_WRL (which we link to from a post) in Search Console shows `
Page is not indexed: Alternative page with proper canonical tag`.

### Options we have

* One way forward could be to simply remove the canonical url from all explorer pages. Then all pages will be crawled separately. But:
	* We don't give the canonical page a higher importance, which we would like to do.
	* Google might decide on a canonical page on its own (because the pages are sufficiently similar), and we can't do something about that.
	* We also index other params, like `?country=` and the like, which we would like to ignore.
* Another way - that we currently cannot support - would be to programmatically set the canonical URL for each view.
	* For example, the url `global-food?facet=none&Food=Kiwi&Metric=Production&Per+Capita=false&country=USA~DEU~FRA~GBR~BRA~ZAF` would result in the canonical url `https://ourworldindata.org/explorers/global-food?facet=none&Food=Kiwi&Metric=Production&Per+Capita=false`, stripping out the `&country=` part.
	* We would need at least some server-side rendering for that to work.
	* Or is it maybe possible to alter the `canonicalUrl` in JS? I'll try that.

**Comment by marcelgerber** [comment]
We have a few constraints that we have to deal with in this project:
* We can only really experiment on the live site, since we don't want staging sites to be findable in Google Search.
* "Experimenting" really means changing some things, waiting for at least a day, and then seeing what happens 🤷🏻
  I would hope that including a new entry in the sitemap leads to swift indexing, but Google also only checks the sitemap once a day so there's definitely quite a bit of patience involved.

Given all of the above, I propose the following experiment:
* Introduce a new explorer property, `indexViewsSeparately` (default false), which controls whether an explorer is indexed with every view or only with the one canonical view.
    * This affects what is included in the sitemap, and how the `canonicalUrl` behaves. Details below ⬇️
* Enable this property on one or two explorers:
    * The [Fossil Fuels Data Explorer](https://ourworldindata.org/explorers/natural-resources) seems like a good fit, because it
        * is not super popular, so if we end up downranking it we don't shoot ourselves in the foot.
        * is _not_ composed of any `grapherId` charts, so searching for a chart title contained in the explorer cannot lead to a normal chart of the same name.
        * contains lots of views for which we don't have a normal charts.
        * only contains 38 views.
    * The [Water and sanitation explorer](https://ourworldindata.org/explorers/water-and-sanitation) (90 different views) is also an option, for similar reasons.
* Deploy it like this and see what happens ⏳
  In particular, we want to see if specific queries like `net imports of coal per capita site:ourworldindata.org` yield specific explorer views as a result.

---

<details>
<summary><code>indexViewsSeparately</code> details</summary>

`indexViewsSeparately`, when enabled, will have the following effects:
* In the sitemap, we list every single view of the explorer. For example, for the `natural-resources` explorer, we would include
    * `natural-resources?Resource=Oil&Metric=Reserves&Count=Total`
    * `natural-resources?Resource=Coal&Metric=Production&Count=Per+capita`
    * ... and all the other 36 combinations.
* The `<link rel="canonical">` tag will be inserted at runtime. It will be derived from the full page URL, so e.g. `natural-resources?facet=none&Resource=Coal&Metric=Production&Count=Per+capita&country=USA~RUS~NOR~GBR~DEU~SWE~CHN` would be trimmed down to `natural-resources?Resource=Coal&Metric=Production&Count=Per+capita`. This makes sure the canonical URL always matches the URL given in the sitemap.

When the option is disabled, both will behave just as today.
</details>

**Comment by larsyencken** [comment]
@marcelgerber Good plan of attack! Let's do it on both of these explorers and see how Google's indexer responds.

**Comment by marcelgerber** [comment]
`natural-resources` and `water-and-sanitation` are now being exposed to Google with all views.
Now it's wait-and-see for us.

**Comment by larsyencken** [comment]
Good news! We have results back:

- Google is indexing all the different configurations
- But... since their page metadata is too similar, Google omits them as near-duplicate results

https://user-images.githubusercontent.com/36012/190622620-57fd8228-01f0-4496-a0d4-343b85fc6091.mov

Seems like we'd want to change the page title for all but the default explorer view as a next step here.

**Comment by larsyencken** [comment]
We can create a new issue for it.

**Comment by marcelgerber** [comment]
Update: After #1661, a very tiny set of explorer views is available in Google Search with the explorer title: [site:ourworldindata.org/explorers Share of the rural population with access to basic sanitation](https://www.google.com/search?q=site%3Aourworldindata.org%2Fexplorers+Share+of+the+rural+population+with+access+to+basic+sanitation)
![image](https://user-images.githubusercontent.com/2641501/191578245-4f9ffc98-aad4-45f5-9a93-e6cc63f77000.png)

It seems it's currently limited to only 1 or 2 pages that have been updated, even though we set the `sitemap.xml` `lastmod` property. It might be the case that Google throttles crawling pages that have a large time overhead, which our explorer pages definitely do.
